# HANDS-ON EXERCISE 1
## Business Analytics — Sales Performance Calculator

In [1]:
import numpy as np

In [2]:
# Task 1: Create the Dataset
daily_sales = np.array([125000, 150000, 175000, 140000, 190000, 210000, 160000])
print("Daily Sales:", daily_sales)
print("Type:", type(daily_sales))
print("Shape:", daily_sales.shape)

print()

# Task 2: Calculate Total Sales
total_sales = np.sum(daily_sales)
print("Total Sales for the Week: ₦", total_sales)

print()

# Task 3: Calculate Average Daily Sales
average_sales = np.mean(daily_sales)
print("Average Daily Sales: ₦", average_sales)
print("Average Daily Sales: ₦", round(average_sales, 2))

print()

# Task 4: Apply a 10% Sales Adjustment
# adjusted_sales = daily_sales * (1 + 10/100)
adjusted_sales = daily_sales * 1.10
print("Adjusted Sales:", adjusted_sales)

print()

# Task 5: Compare Original and Adjusted Sales
additional_revenue = adjusted_sales - daily_sales
print("Additional Revenue per Day:", additional_revenue)

print()

# Task 6: Use Numpy's Mathematical Functions(Compare Average and Total Sales)
print("Average Daily Sales: ₦", round(average_sales, 2))
print("Total Weekly Sales: ₦", total_sales)

Daily Sales: [125000 150000 175000 140000 190000 210000 160000]
Type: <class 'numpy.ndarray'>
Shape: (7,)

Total Sales for the Week: ₦ 1150000

Average Daily Sales: ₦ 164285.7142857143
Average Daily Sales: ₦ 164285.71

Adjusted Sales: [137500. 165000. 192500. 154000. 209000. 231000. 176000.]

Additional Revenue per Day: [12500. 15000. 17500. 14000. 19000. 21000. 16000.]

Average Daily Sales: ₦ 164285.71
Total Weekly Sales: ₦ 1150000


## Average Daily Sales vs Total Sales 
The total sales figure shows the overall revenue generated by the business during the week, while the average daily sales shows the typical amount of revenue generated per day. The average helps management understand daily sales performance and provides a useful benchmark for comparing individual days.

# HANDS-ON EXERCISE 2
## Education Analytics — Student Performance Analysis


In [3]:
# Task 1: Create and Inspect the Data
scores = np.array([62, 75, 81, 69, 88, 94, 73, 85, 77, 91])
print("Scores:", scores)
print("Type:", type(scores))
print("Shape:", scores.shape)

print()


# Task 2: Calculate the Mean Score
mean_score = np.mean(scores)
print("Mean Score:", mean_score)

print()


# Task 3: Calculate Each Student's Deviation
deviations = scores - mean_score
print("Deviations:", deviations)

print()


# Task 4: Identify Positive and Negative Deviations (Understanding Deviations)
# A positive deviation means that a student's score is above the group average.
# A negative deviation means that a student's score is below the group average.
# A deviation close to zero means that the student's score is close to the group average.

print()


# Task 5: Square the Deviations
squared_deviations = deviations ** 2
print("Squared Deviations:", squared_deviations)


print()


# Task 6: Calculate the Sum of Squared Deviations
sum_squared_deviations = np.sum(squared_deviations)
print("Sum of Squared Deviations:", sum_squared_deviations)

Scores: [62 75 81 69 88 94 73 85 77 91]
Type: <class 'numpy.ndarray'>
Shape: (10,)

Mean Score: 79.5

Deviations: [-17.5  -4.5   1.5 -10.5   8.5  14.5  -6.5   5.5  -2.5  11.5]


Squared Deviations: [306.25  20.25   2.25 110.25  72.25 210.25  42.25  30.25   6.25 132.25]

Sum of Squared Deviations: 932.5


## Why Do We Square the Deviations?
Squaring the deviations changes all negative values into positive values because multiplying a negative number by itself produces a positive number. This prevents positive and negative deviations from cancelling each other out when we calculate the overall variation in the scores.

## Standard Deviation Process
The standard deviation calculation can be understood through these steps:

1. Calculate the mean.
2. Calculate the deviation of each score from the mean.
3. Square the deviations.
4. Sum the squared deviations.
5. Divide by the appropriate number of observations.
6. Take the square root.

Standard deviation is useful because it shows how spread out the students' scores are around the average. While the mean tells us the typical score, standard deviation gives us additional information about whether the scores are closely grouped around the mean or widely spread out.

## Industry Thinking Challenge
The average score gives an overall picture of student performance, but it does not show how individual scores are distributed around that average. Looking at deviations and standard deviation helps the coordinator identify whether students are performing at similar levels or whether there are large differences in performance that may require further attention.

# HANDS-ON EXERCISE 3
## Engineering & Scientific Computing — Trigonometric Series and Convergence

In [4]:
# Task 1: Define the angle
theta = 30
theta_radians = np.radians(theta)
print("Theta in degrees:", theta)
print("Theta in radians:", theta_radians)

print()


# Task 2: Create the sequence
k = np.arange(1, 100, 1)
print("First value:", k[0])
print("Last value:", k[-1])
print("Number of values:", len(k))

print()


# Task 3: Calculate the trigonometric component
sin_theta = np.sin(theta_radians)
print("sin(theta):", sin_theta)

print()


# Task 4: Build the series terms
series_terms = sin_theta / k
print("Series Terms:", series_terms)

print()


# Task 5: Calculate the sum
series_sum = np.sum(series_terms)
print("Series Sum:", series_sum)

print()


# Task 6: Increase the number of terms
k_99 = np.arange(1, 100, 1)
series_99 = np.sum(sin_theta / k_99)

k_999 = np.arange(1, 1000, 1)
series_999 = np.sum(sin_theta / k_999)

k_9999 = np.arange(1, 10000, 1)
series_9999 = np.sum(sin_theta / k_9999)

print("99 terms:", series_99)
print("999 terms:", series_999)
print("9,999 terms:", series_9999) 

Theta in degrees: 30
Theta in radians: 0.5235987755982988

First value: 1
Last value: 99
Number of values: 99

sin(theta): 0.49999999999999994

Series Terms: [0.5        0.25       0.16666667 0.125      0.1        0.08333333
 0.07142857 0.0625     0.05555556 0.05       0.04545455 0.04166667
 0.03846154 0.03571429 0.03333333 0.03125    0.02941176 0.02777778
 0.02631579 0.025      0.02380952 0.02272727 0.02173913 0.02083333
 0.02       0.01923077 0.01851852 0.01785714 0.01724138 0.01666667
 0.01612903 0.015625   0.01515152 0.01470588 0.01428571 0.01388889
 0.01351351 0.01315789 0.01282051 0.0125     0.01219512 0.01190476
 0.01162791 0.01136364 0.01111111 0.01086957 0.0106383  0.01041667
 0.01020408 0.01       0.00980392 0.00961538 0.00943396 0.00925926
 0.00909091 0.00892857 0.00877193 0.00862069 0.00847458 0.00833333
 0.00819672 0.00806452 0.00793651 0.0078125  0.00769231 0.00757576
 0.00746269 0.00735294 0.00724638 0.00714286 0.00704225 0.00694444
 0.00684932 0.00675676 0.00666667 0.00

# Task 7: Investigate Convergence
| Number of Terms | Series Result |
|---:|---:|
| 99 | 2.58869 |
| 999 | 3.74224 |
| 9,999 | 4.89375 |
The result increases as more terms are added

# Task 8: Understanding Convergence
In this experiment, increasing the number of terms causes the series result to increase. With 99 terms, the result is approximately 2.58869, while 999 terms gives approximately 3.74224 and 9,999 terms gives approximately 4.89375.

For a convergent series, increasing the number of terms would cause the result to approach a fixed limiting value. However, the series used in this exercise is proportional to the harmonic series, which does not converge. Therefore, the result continues to increase as more terms are added rather than approaching a fixed limit.

# BONUS CHALLENGE- Food Production Performance Analysis

### Food Production Performance Analysis Using NumPy
This mini-project uses NumPy to analyze daily food production output over one week.

### Scenario
A food processing company records the number of food packages produced each day.
The production manager wants to understand the company's weekly production,
average daily output, and how much production would increase if output improved
by 10%.

### Objective

The objective of this experiment is to use NumPy to calculate the total and
average production, apply a 10% production increase, and determine the
additional number of packages that could be produced.

### Dataset
The following dataset represents the number of food packages produced by the
company over seven days. The data is simulated for this numerical experiment.

In [7]:
# Dataset
daily_production = np.array([1200, 1350, 1180, 1450, 1500, 1320, 1250])

print()

# Inspect the array
print("Daily Production:", daily_production)
print("Type:", type(daily_production))
print("Shape:", daily_production.shape)

print()

# Calculate total production
total_production = np.sum(daily_production)
print("Total Weekly Production:", total_production)

print()

# Calculate average production
average_production = np.mean(daily_production)
print("Average Daily Production:", round(average_production, 2))

print()

# Apply 10% increase
adjusted_production = daily_production * (1 + 10/100)
print("Adjusted Production:", adjusted_production)

print()

# Calculate additional production
additional_production = adjusted_production - daily_production
print("Additional Production:", additional_production)

print()

# Calculate total adjusted production
total_adjusted_production = np.sum(adjusted_production)
print("Original Weekly Production:", total_production)
print("Adjusted Weekly Production:", total_adjusted_production)


Daily Production: [1200 1350 1180 1450 1500 1320 1250]
Type: <class 'numpy.ndarray'>
Shape: (7,)

Total Weekly Production: 9250

Average Daily Production: 1321.43

Adjusted Production: [1320. 1485. 1298. 1595. 1650. 1452. 1375.]

Additional Production: [120. 135. 118. 145. 150. 132. 125.]

Original Weekly Production: 9250
Adjusted Weekly Production: 10175.0


### Results

| Measurement | Result |
|---|---:|
| Total weekly production | 9,250 packages |
| Average daily production | 1,321.43 packages |
| Production after 10% increase | 10,175 packages |
| Additional weekly production | 925 packages |

### Interpretation

The company produced a total of 9,250 packages during the week, with an average
daily production of approximately 1,321 packages. A hypothetical 10% increase
in production would raise weekly output to 10,175 packages, giving the company
925 additional packages for the week.

The average daily production provides a useful benchmark for understanding the
company's normal daily output, while the total production shows the overall
weekly output. The adjusted figures also help management estimate the potential
effect of improving production efficiency.

### Conclusion

NumPy can be used to quickly analyze production data and perform calculations
on an entire array at once. This experiment demonstrated how numerical tools
can help a food production company understand its output and estimate the
potential impact of increased production.